docs/docs/env-variables.md
# Environmant Variables

### Gerneral

| Variable                                                          | Default | Description                                              |
| :---------------------------------------------------------------- | :-----: | :------------------------------------------------------- |
| BASE_URL :fontawesome-solid-exclamation-circle:{title="required"} |         |                                                          |
| REQUIRE_SIGNUP_CODE                                               |  true   | Allow use sigup without token                            |
| SIGNUP_CODE_LIFETIME                                              |   24    | The time in hours a signup code is usbale                |
| EMAIL_REQUIRE_VERIFICATION                                        |  false  | Require users to click a varification link in an E-Mail. |
| ENVIRONMENT                                                       |  prod   |                                                          |

### Security

| Variable                                                            | Default | Description                        |
| :------------------------------------------------------------------ | :-----: | :--------------------------------- |
| ACCESS_TOKEN_EXPIRATION_HOURS                                       |   192   | The time in hours a token is valid |
| SECRET_KEY :fontawesome-solid-exclamation-circle:{title="required"} |         |                                    |

### Database

| Variable                                                                   |  Default   | Description |
| :------------------------------------------------------------------------- | :--------: | :---------- |
| POSTGRES_SERVER :fontawesome-solid-exclamation-circle:{title="required"}   |            |             |
| POSTGRES_PORT                                                              |    5432    |             |
| POSTGRES_USER :fontawesome-solid-exclamation-circle:{title="required"}     |            |             |
| POSTGRES_PASSWORD :fontawesome-solid-exclamation-circle:{title="required"} |            |             |
| POSTGRES_DB                                                                | nightdrive |             |

### E-Mail

SMTP server setup.

| Variable        | Default | Description |
| :-------------- | :-----: | :---------- |
| SMTP_TLS        |  true   |             |
| SMTP_SSL        |  false  |             |
| SMTP_PORT       |   587   |             |
| SMTP_HOST       |         |             |
| SMTP_USER       |         |             |
| SMTP_PASSWORD   |         |             |
| SMTP_FROM_EMAIL |         |             |
| SMTP_FROM_NAME  |         |             |


docs/docs/deployment/env-variables.md
---
icon: material/variable
description: Overview and explanation of enviroment variables.
---

# Environmant Variables

### Gerneral

| Variable                                                          | Default | Description                                                                                                                |
| :---------------------------------------------------------------- | :-----: | :------------------------------------------------------------------------------------------------------------------------- |
| BASE_URL :fontawesome-solid-exclamation-circle:{title="required"} |         |                                                                                                                            |
| REQUIRE_SIGNUP_CODE                                               |  true   | Require users to provide a code to sign up                                                                                 |
| SIGNUP_CODE_LIFETIME                                              |   24    | The time in hours a signup code is usable                                                                                  |
| EMAIL_REQUIRE_VERIFICATION                                        |  false  | Require users to click a varification link in an E-Mail                                                                    |
| ENVIRONMENT                                                       |  prod   | Operating environment either prod or dev                                                                                   |
| MESSAGE_OF_THE_DAY                                                |  None   | Message to be displayed on login or signup screen                                                                          |
| TIME_ZONE                                                         | Etc/UTC | Time zone identifier from this [list](https://en.wikipedia.org/wiki/List_of_tz_database_time_zones#List){:target="_blank"} |


### Security

| Variable                                                            | Default | Description                        |
| :------------------------------------------------------------------ | :-----: | :--------------------------------- |
| ACCESS_TOKEN_EXPIRATION_HOURS                                       |   192   | The time in hours a token is valid |
| SECRET_KEY :fontawesome-solid-exclamation-circle:{title="required"} |         | Used to generate JWT               |

### Passwords

Password requirement settings.

| Variable                        | Default | Description                                                                      |
| :------------------------------ | :-----: | :------------------------------------------------------------------------------- |
| PASSWORD_MIN_LENGTH             |   12    | Minimum password length (can't be set `<8`)                                      |
| PASSWORD_REQUIRE_NUMBERS        |  True   | Require passwords to include numbers `0-9`                                       |
| PASSWORD_REQUIRE_UPER_LOWERCASE |  True   | Require password to include upper and lowercase letters `a-z`, `A-Z`             |
| PASSWORD_REQUIRE_SYMBOLS        |  True   | Require password to include special characters `~!@#$%^&*()_-+={[}]|\:;"'<,>.?/` |

### Database

PostgreSQL server connection.

| Variable                                                                   |  Default   | Description                      |
| :------------------------------------------------------------------------- | :--------: | :------------------------------- |
| POSTGRES_SERVER :fontawesome-solid-exclamation-circle:{title="required"}   |            | Postgres database server address |
| POSTGRES_PORT                                                              |    5432    | Postgres database port           |
| POSTGRES_USER                                                              | nightdrive | Postgres database user           |
| POSTGRES_PASSWORD :fontawesome-solid-exclamation-circle:{title="required"} |            | Postgres database password       |
| POSTGRES_DB                                                                | nightdrive | Postgres database name           |

### E-Mail

SMTP server setup.

| Variable        |  Default   | Description             |
| :-------------- | :--------: | :---------------------- |
| SMTP_AUTH_TYPE  |    TLS     | Options: `TLS` or `SSL` |
| SMTP_PORT       |    587     |                         |
| SMTP_HOST       |            |                         |
| SMTP_USER       | nightdrive |                         |
| SMTP_PASSWORD   |            |                         |
| SMTP_FROM_EMAIL |            |                         |
| SMTP_FROM_NAME  | NightDrive |                         |


backend/app/core/config.py
import secrets
import warnings
from typing import Annotated, Any, Literal

from pydantic import (
    AnyUrl,
    EmailStr,
    PostgresDsn,
    HttpUrl,
    BeforeValidator,
    computed_field
)

from pydantic_core import MultiHostUrl

from pydantic_settings import BaseSettings, SettingsConfigDict

class Settings(BaseSettings):
    model_config = SettingsConfigDict(
        env_file="../.env",
        env_ignore_empty=True,
        extra="ignore"
    )

    ENVIRONMENT: Literal["dev", "prod"] = "prod"

    LOCALES: str = "en"
    def LOCALES_AVAILABLE(self) -> list:
        return self.LOCALES.split(",")

    PROJECT_NAME: str = "NightDrive"

    MESSAGE_OF_THE_DAY: str = None

    BASE_URL: str = ""
    API_V1_STR: str = "/api/v1"

    TIME_ZONE: str = "Etc/UTC"

    SECRET_KEY: str

    # in days (default: 8)
    ACCESS_TOKEN_EXPIRATION_HOURS: int = 24*8

    # in hours (default: 24)
    REQUIRE_SIGNUP_CODE: bool = True
    SIGNUP_CODE_LIFETIME: int = 24
    """Hours a signup code is valid for."""

    ### Password settings
    PASSWORD_MIN_LENGTH: int = 12
    PASSWORD_REQUIRE_NUMBERS: bool = True
    PASSWORD_REQUIRE_UPER_LOWERCASE: bool = True
    PASSWORD_REQUIRE_SYMBOLS: bool = True
    PASSWORD_SYMBOLS_LIST: str = "~!@#$%^&*()_-+={[}]|\:;\"\'<,>.?/"

    ### PostgreSQL database settings    
    POSTGRES_SERVER: str
    POSTGRES_PORT: int = 5432
    POSTGRES_USER: str = "nightdrive"
    POSTGRES_PASSWORD: str
    POSTGRES_DB: str = "nightdrive"

    @computed_field
    @property
    def SQLALCHEMY_DATABASE_URI(self) -> PostgresDsn:
        return MultiHostUrl.build(
            scheme="postgresql+psycopg",
            username=self.POSTGRES_USER,
            password=self.POSTGRES_PASSWORD,
            host=self.POSTGRES_SERVER,
            port=self.POSTGRES_PORT,
            path=self.POSTGRES_DB
        )


    ### E-Mail settings  
    EMAIL_REQUIRE_VERIFICATION: bool = False

    SMTP_AUTH_TYPE: Literal["TLS", "SSL"] = "TLS"
    SMTP_SSL: bool = False
    SMTP_PORT: int = 587
    SMTP_HOST: str | None = None
    SMTP_USER: str | None = None
    SMTP_PASSWORD: str | None = None
    EMAIL_FROM_EMAIL: EmailStr | None = None
    EMAIL_FROM_NAME: EmailStr | None = None

    @computed_field
    @property
    def EMAIL_CONFIGURED(self) -> bool:
        return bool(self.SMTP_HOST and self.EMAIL_FROM_EMAIL)
    

    ### Live group settings
    LIVE_GROUP_MAX_DURATION: int = 24*60
    """Maximum duartion of a life group in Minutes."""

settings = Settings()
""" NightDrive API configuration values. """